toolchain/check/context.cpp

```diff
@@ -421,13 +421,14 @@ auto Context::LookupUnqualifiedName(Parse::NodeId node_id,
           .scope_result = SemIR::ScopeLookupResult::MakeError()};
 }
 
-auto Context::LookupNameInExactScope(SemIRLoc loc, SemIR::NameId name_id,
+auto Context::LookupNameInExactScope(SemIR::LocId loc_id, SemIR::NameId name_id,
                                      SemIR::NameScopeId scope_id,
                                      SemIR::NameScope& scope,
                                      bool is_being_declared)
     -> SemIR::ScopeLookupResult {
-  if (auto entry_id = is_being_declared ? scope.Lookup(name_id)
-                                        : scope.LookupOrPoison(name_id)) {
+  if (auto entry_id = is_being_declared
+                          ? scope.Lookup(name_id)
+                          : scope.LookupOrPoison(loc_id, name_id)) {
```

Comment:
In order to pass a `LocId` here, you're changing `impl.cpp` to pass one in instead of a `SemIRLoc`. However, because `is_being_declared` is always false there, this creates a frequently unnecessary lookup of the `LocId`.

Had you considered splitting apart the `is_being_declared = true` portion that requires a `LocId` from the alternative that only uses the loc with `ImportNameFromOtherPackage` (which might not diagnose), in order to reduce fetches?

Reply:
IIUC, when `is_being_declared` is false, `LocId` is necessary, so the lookup is required.
The only case where `is_being_declared` is true already has a `LocId`.
To avoid the extra lookup, we will need to allow `SemIRLoc` or something similar in `NameScope`.